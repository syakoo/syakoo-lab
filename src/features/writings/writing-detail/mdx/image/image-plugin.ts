import sharp from "sharp";

import type { MDXCustomTextPlugin } from "@/features/mdx/types";

// replace の async 版
const asyncReplace = async (
  originalText: string,
  pattern: RegExp,
  replacer: (match: RegExpMatchArray) => Promise<string>,
) => {
  const matchedTexts = [...originalText.matchAll(pattern)];
  if (!matchedTexts) return originalText;

  const replacePatterns = await Promise.all(
    matchedTexts.map(async (match) => {
      match;
      const result = await replacer(match);
      return [match[0], result] as [string, string];
    }),
  );

  const replacedText = replacePatterns.reduce((prev, replacePattern) => {
    return prev.replace(replacePattern[0], replacePattern[1]);
  }, originalText);

  return replacedText;
};

export const imagePlugin: MDXCustomTextPlugin = async (mdxText) => {
  const pattern = /<Image([\s\S]*?)\/>/g;

  return await asyncReplace(mdxText, pattern, async ([match, attributes]) => {
    const srcMatch = attributes.match(/src={?["'](.*?)["']/);
    // NOTE: src がマッチしない或いはこのサイトのものではない場合はそのまま返却する
    if (!srcMatch || !srcMatch[1].startsWith("/")) {
      return match;
    }
    const imageUrl = `${process.cwd()}/public${srcMatch[1]}`;
    const { width, height } = await sharp(imageUrl).metadata();

    return `<Image${attributes}\n width={${width}} height={${height}} \n/>`;
  });
};
