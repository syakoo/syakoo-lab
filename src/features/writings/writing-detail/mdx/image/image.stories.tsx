import type { Meta, StoryObj } from "@storybook/nextjs";

import { Image } from ".";

const meta = {
  component: Image,
  parameters: {},
} satisfies Meta<typeof Image>;

export default meta;

type Story = StoryObj<typeof Image>;

export const Sample: Story = {
  args: {
    caption: "Sample Caption",
    src: "https://placehold.jp/3063D4/ffffff/400x300.png",
    width: 400,
    height: 300,
  },
};
