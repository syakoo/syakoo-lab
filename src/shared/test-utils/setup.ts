import { exampleMDXComponent } from "@/features/mdx/mocks/fixture-mdxcomponent";

import { loadEnv } from "./nextjs/env";

// NOTE: `@mdx-js/mdx` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
jest.mock("@mdx-js/mdx", () => {
  return {
    runSync: jest
      .fn()
      .mockImplementation(() => ({ default: exampleMDXComponent })),
  };
});

// NOTE: `mermaid` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
jest.mock("mermaid", () => ({
  initialize: () => {
    return;
  },
  run: () => {
    return;
  },
}));

loadEnv();

jest.spyOn(Math, "random").mockImplementation(() => 0.5866245602626035);
