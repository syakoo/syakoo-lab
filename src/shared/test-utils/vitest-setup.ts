import { vi } from "vitest";

import { exampleMDXComponent } from "@/features/mdx/mocks/fixture-mdxcomponent";

import { loadEnv } from "./nextjs/env";

// NOTE: Next.jsのImageコンポーネントをテスト用にモック
vi.mock("next/image", () => ({
  default: "img",
}));

// NOTE: `@mdx-js/mdx` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
vi.mock("@mdx-js/mdx", () => {
  return {
    runSync: vi
      .fn()
      .mockImplementation(() => ({ default: exampleMDXComponent })),
  };
});

// NOTE: `mermaid` の esm がうまく解決できなかったためまとめてモック化
// これが原因でおかしくなっていたらごめんなさい
vi.mock("mermaid", () => ({
  default: {
    initialize: () => {
      return;
    },
    run: () => {
      return;
    },
  },
  initialize: () => {
    return;
  },
  run: () => {
    return;
  },
}));

loadEnv();

vi.spyOn(Math, "random").mockImplementation(() => 0.5866245602626035);
