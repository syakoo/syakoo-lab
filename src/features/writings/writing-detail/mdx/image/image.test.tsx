import { testStories } from "@/shared/test-utils/test-stories";

import * as stories from "./image.stories";

describe("Image", () => {
  testStories(stories);
});
