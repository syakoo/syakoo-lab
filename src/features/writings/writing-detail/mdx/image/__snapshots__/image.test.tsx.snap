// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image > スナップショット Sample 1`] = `
<body>
  <div>
    <figure
      class="image_imageStyles_figure__c4o45q0"
    >
      <img
        alt="Sample Caption"
        class="image_imageStyles_image__c4o45q2"
        height="300"
        src="https://placehold.jp/3063D4/ffffff/400x300.png"
        width="400"
      />
      <figcaption
        class="image_imageStyles_caption__c4o45q1"
      >
        Sample Caption
      </figcaption>
    </figure>
  </div>
</body>
`;
